File: fastapi-python-backend/src/routes/health_route.py
import time
from datetime import UTC, datetime

from fastapi import APIRouter

from ..config.env import settings
from ..errors.http_errors import ServiceUnavailableError
from ..state.readiness import is_service_ready

_PROCESS_START_TIME = time.monotonic()

health_router = APIRouter(prefix='/health', tags=['health'])
ready_router = APIRouter(prefix='/ready', tags=['health'])


def _utc_now_iso() -> str:
    return datetime.now(UTC).isoformat().replace('+00:00', 'Z')


@health_router.get('')
async def get_health() -> dict[str, object]:
    return {
        'success': True,
        'data': {
            'status': 'ok',
            'uptime': time.monotonic() - _PROCESS_START_TIME,
            'timestamp': _utc_now_iso(),
            'environment': settings.NODE_ENV,
        },
    }


@ready_router.get('')
async def get_ready() -> dict[str, object]:
    if not is_service_ready():
        raise ServiceUnavailableError(
            detail='Service is shutting down and not ready to serve traffic.'
        )

    return {
        'success': True,
        'data': {
            'ready': True,
            'timestamp': _utc_now_iso(),
        },
    }

File: express-ts-bun-backend/src/routes/health.route.ts
import { Router } from 'express';
import { env } from '../config/env';
import { ServiceUnavailableError } from '../errors/http-errors';
import { isServiceReady } from '../state/readiness';

const healthRouter = Router();
const readyRouter = Router();

healthRouter.get('/', (_req, res) => {
  res.status(200).json({
    success: true,
    data: {
      status: 'ok',
      uptime: process.uptime(),
      timestamp: new Date().toISOString(),
      environment: env.NODE_ENV,
    },
  });
});

readyRouter.get('/', (_req, res, next) => {
  if (!isServiceReady()) {
    next(
      new ServiceUnavailableError({
        detail: 'Service is shutting down and not ready to serve traffic.',
      }),
    );
    return;
  }

  res.status(200).json({
    success: true,
    data: {
      ready: true,
      timestamp: new Date().toISOString(),
    },
  });
});

export { healthRouter, readyRouter };
